File: Chat/Server/src/Chat/Script/CsScript/Action/Action3002.cs
using System;
using GameServer.Script.CsScript.Com;
using GameServer.Script.CsScript.JsonProtocol;
using GameServer.Script.Model;
using ZyGames.Framework.Game.Service;
using System.Collections.Generic;
using ZyGames.Framework.Game.Model;
using ZyGames.Framework.Cache.Generic;

namespace GameServer.Script.CsScript.Action
{
    /// <summary>
    /// chat list
    /// </summary>
    public class Action3002 : BaseAction
    {
        private List<ChatMessage> chatList;

        public Action3002(ActionGetter actionGetter)
            : base((short)3002, actionGetter)
        {

        }

        /// <summary>
        /// 客户端请求的参数较验
        /// </summary>
        /// <returns>false:中断后面的方式执行并返回Error</returns>
        public override bool GetUrlElement()
        {
            return true;
        }

        /// <summary>
        /// 业务逻辑处理
        /// </summary>
        /// <returns>false:中断后面的方式执行并返回Error</returns>
        public override bool TakeAction()
        {
            var chatservice = new MyChatService(Current.User as ChatUser);
            chatList = chatservice.PopMessages();
            return true;
        }


        protected override string BuildResponsePack()
        {
            var result = new ChatData[chatList.Count];
            for (int i = 0; i < chatList.Count; i++)
            {
                var chat = chatList[i];
                result[i] = new ChatData()
                {
                    Id = chat.Version,
                    UserId = chat.FromUserID,
                    Message = chat.Content,
                    Sender = GetUserName(chat.FromUserID),
                    SendTime = chat.SendDate.ToString("HH:mm:ss")
                };
            }
            Body = result;
            return base.BuildResponsePack();
        }

        private string GetUserName(int userId)
        {
            var cache = new MemoryCacheStruct<ChatUser>();
            ChatUser user;
            if (cache.TryGet(userId.ToString(), out user))
            {
                return user.UserName;
            }
            return string.Empty;
        }

    }
}

File: Chat/Server/src/Chat/Script/PyScript/Action/action3002.py
"""chat list"""
import clr, sys
import json
import ReferenceLib
from action import *
from System import *
from ZyGames.Framework.Common import *
from ZyGames.Framework.Game.Contract import *
from ZyGames.Framework.Game.Service import *
from ZyGames.Framework.RPC.Sockets import *
from ZyGames.Framework.Cache.Generic import *
from GameServer.Script.CsScript.Com import *
from GameServer.Script.Model import *
#引用示例
#clr.AddReference('ZyGames.Framework.Common');
#from ZyGames.Framework.Common.Log import *

class ChatData():

    def __init__(self):
        self.Id = 0
        self.Message = ''
        self.UserId = 0
        self.Sender = ''
        self.SendTime = ''
        return

class UrlParam(HttpParam):
    def __init__(self):
        HttpParam.__init__(self)


class ActionResult(DataResult):
    def __init__(self):
        DataResult.__init__(self)


def getUrlElement(httpGet, parent):
    urlParam = UrlParam()
    urlParam.MsgId = httpGet.GetMsgId()
    urlParam.ActionId = httpGet.GetActionId()
    return urlParam

def takeAction(urlParam, parent):
    actionResult = ActionResult()
    user = parent.Current.User
    chatservice = MyChatService(user)
    actionResult.chatList = chatservice.PopMessages()
    return actionResult

def buildPacket(urlParam, actionResult):
    list = []
    for chat in actionResult.chatList:
        data = ChatData()
        data.Id = chat.Version
        data.UserId = chat.FromUserID
        data.Message = chat.Content
        data.Sender = GetUserName(chat.FromUserID)
        data.SendTime = chat.SendDate.ToString("HH:mm:ss")
        list.append(data)

    result = JsonDataResult(urlParam)
    result.setBody(list)
    jsonStr = json.dumps(result, default=convert_to_builtin_type)
    return jsonStr


def GetUserName(userId):
    cache = MemoryCacheStruct[ChatUser]()
    result = cache.TryGet(str(userId))
    if result[0] == True:
        return result[1].UserName
    else:
        return ''

def convert_to_builtin_type(obj): 
    d = {  }
    d.update(obj.__dict__)
    return d